pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,253 @@
+		cnpCpy := cnp.DeepCopy()
+
+		translationStart := time.Now()
+		translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
+		metrics.CIDRGroupTranslationTimeStats.Observe(time.Since(translationStart).Seconds())
+
+		err := k.updateCiliumNetworkPolicyV2(cs, cnpCpy, translatedCNP, initialRecvTime)
+		if err == nil {
+			cnpCache[key] = cnpCpy
+		}
+
+		errs = append(errs, err)
+	}
+	return errors.Join(errs...)
+}
+
+func resolveCIDRGroupRef(cnp *types.SlimCNP, cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup) *types.SlimCNP {
+	refs := getCIDRGroupRefs(cnp)
+	if len(refs) == 0 {
+		return cnp
+	}
+
+	cidrsSets, err := cidrGroupRefsToCIDRsSets(refs, cidrGroupCache)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			logfields.K8sAPIVersion:           cnp.TypeMeta.APIVersion,
+			logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+			logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+			logfields.CIDRGroupRefs:           refs,
+		}).WithError(err).Warning("unable to translate all cidr groups to cidrs")
+	}
+
+	translated := translateCIDRGroupRefs(cnp, cidrsSets)
+
+	return translated
+}
+
+func hasCIDRGroupRef(cnp *types.SlimCNP, cidrGroup string) bool {
+	if specHasCIDRGroupRef(cnp.Spec, cidrGroup) {
+		return true
+	}
+	for _, spec := range cnp.Specs {
+		if specHasCIDRGroupRef(spec, cidrGroup) {
+			return true
+		}
+	}
+	return false
+}
+
+func specHasCIDRGroupRef(spec *api.Rule, cidrGroup string) bool {
+	if spec == nil {
+		return false
+	}
+	for _, ingress := range spec.Ingress {
+		for _, rule := range ingress.FromCIDRSet {
+			if string(rule.CIDRGroupRef) == cidrGroup {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getCIDRGroupRefs(cnp *types.SlimCNP) []string {
+	var cidrGroupRefs []string
+
+	if cnp.Spec != nil {
+		for _, ingress := range cnp.Spec.Ingress {
+			for _, rule := range ingress.FromCIDRSet {
+				cidrGroupRefs = append(cidrGroupRefs, string(rule.CIDRGroupRef))
+			}
+		}
+	}
+	for _, spec := range cnp.Specs {
+		for _, ingress := range spec.Ingress {
+			for _, rule := range ingress.FromCIDRSet {
+				cidrGroupRefs = append(cidrGroupRefs, string(rule.CIDRGroupRef))
+			}
+		}
+	}
+
+	return cidrGroupRefs
+}
+
+func cidrGroupRefsToCIDRsSets(cidrGroupRefs []string, cache map[string]*cilium_v2_alpha1.CiliumCIDRGroup) (map[string][]api.CIDR, error) {
+	var errs []error
+	cidrsSet := make(map[string][]api.CIDR)
+	for _, cidrGroupRef := range cidrGroupRefs {
+		var found bool
+		for cidrGroupName, cidrGroup := range cache {
+			if cidrGroupName == cidrGroupRef {
+				cidrs := make([]api.CIDR, 0, len(cidrGroup.Spec.ExternalCIDRs))
+				for _, cidr := range cidrGroup.Spec.ExternalCIDRs {
+					cidrs = append(cidrs, api.CIDR(cidr))
+				}
+				cidrsSet[cidrGroupRef] = cidrs
+				found = true
+				break
+			}
+		}
+		if !found {
+			errs = append(errs, fmt.Errorf("cidr group %s not found, skipping translation", cidrGroupRef))
```

Comment:
nit: The user input should probably be quoted to make it clear that this is user input and not part of the static message.
```suggestion
			errs = append(errs, fmt.Errorf("cidr group %q not found, skipping translation", cidrGroupRef))
```